### PR TITLE
traceparent restart should require tracestate cleanup rationale

### LINKED ISSUE #146: How and when are you allowed to restart a trace
We discussed the need and option to restart a trace.  The question came up during test case generation. 

Key questions are: 
- When are you allowed to restart a trace?
- Do you also clear ```tracestate``` when you restart?
- Are there multiple types of restart?
  - Soft restart keeping tracestate
  - Hard restart not keeping tracestate

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
Most of the text in this PR is for rationale, not specification. I marked it as editorial as this is what was assumed.

Hopefully this will close #146 and #252

### CODE REVIEW

File: spec/20-http_header_format.md [hunk 1]
```diff
@@ -453,7 +453,10 @@ Here is the list of allowed mutations:
 4. **Restarting trace**. All properties - `trace-id`, `parent-id`, `trace-flags`
    are regenerated. This mutation is used in the services defined as a front
    gate into secure networks and eliminates a potential denial of service attack
-   surface.
+   surface. Implementation SHOULD clean up `tracestate` collection on
+   `traceparent` restart. In rare cases when original `tracestate` entries must
```
Comment (AloisReitbauer) [inline spec/20-http_header_format.md]: article missing: the original

Comment (AloisReitbauer) [inline spec/20-http_header_format.md]: Be specific here. In which rare cases. Without examples adding this does not help the reader to decide.

**Review by AloisReitbauer — CHANGES_REQUESTED** [review]

**Comment by SergeyKanzhelev** [comment]
@AloisReitbauer can you please review again?

File: spec/20-http_header_format.md [hunk 2]
```diff
@@ -453,7 +453,13 @@ Here is the list of allowed mutations:
 4. **Restarting trace**. All properties - `trace-id`, `parent-id`, `trace-flags`
    are regenerated. This mutation is used in the services defined as a front
    gate into secure networks and eliminates a potential denial of service attack
-   surface.
+   surface. Implementation SHOULD clean up `tracestate` collection on
```
Comment (mtwo) [inline spec/20-http_header_format.md]: Implementation -> Implementations

File: spec/21-http_header_format_rationale.md [hunk 3]
```diff
@@ -152,6 +152,34 @@ One variation is whether original or new header that you cannot recognize is pre
...
+than restored back when it is leaving those boundaries - keeping original
+`tracestate` entries will increase tracing vendors interoperability.
+Vendor-specific context will be propagated thru the these secure boundaries.
+
+Scenarios like one above require careful coding and understanding of what they
+are trying to achieve. It should be considered an exception. Implementations
+should discourage this type of restarts. For example, implementations may allow
+this scenario only by means of restarting the trace WITH `tracestate` clean up
+and than re-population of `tracestate` can only be implemented as an explicit
+copy/pasting of `tracestate` entries one by one.
```
Comment (mtwo) [inline spec/21-http_header_format_rationale.md:263]: Modified for grammar / clarity

Developers sometimes feel a need to restart a trace due to security concerns,
however trace context should not contain information that can compromise your
application,  and tracing vendors should work through the risks of trusting
incoming identifiers without compromising on interoperability.

If restarting a trace is unavoidable, you SHOULD clear the `tracestate` list as well.
The data carried in `tracestate` carries information that often has a strong
association with the `traceparent`, particularly `trace-id`. Tracing vendors will
often assume that these values are associated, which is why both `tracestate`
SHOULD be cleared when `traceparent` is changed.

There are scenarios, however, which may require different behavior.
For example, if a trace was restarted when entered some secure boundaries and
than restored back when it is leaving those boundaries - keeping the original
`tracestate` entries will fully restore the trace back to normal. In this example,
vendor-specific context will be propagated through these secure boundaries.

Scenarios like the one above require careful coding and understanding of what they
are trying to achieve, and should be considered an exception. Implementations
should discourage this type of restarts. For example, implementations may allow
this scenario only by means of restarting the trace WITH `tracestate` clean up
and than re-population of `tracestate` can only be implemented as an explicit
copy/pasting of `tracestate` entries one by one.

**Review by mtwo — COMMENTED** [review]
Made a suggestion for grammar / clarity

File: spec/20-http_header_format.md (hunk 2 above)
Comment (SergeyKanzhelev) [reply spec/20-http_header_format.md]: Also fixed the same accross the entire doc

File: spec/21-http_header_format_rationale.md (hunk 3 above)
Comment (SergeyKanzhelev) [reply spec/21-http_header_format_rationale.md:263]: thanks!

File: spec/20-http_header_format.md [hunk 4]
```diff
@@ -242,11 +242,11 @@ base16(trace-flags) = 00  // not recorded
 
 ## Versioning of `traceparent`
 
-Implementation is opinionated about future version of the specification. Current
+Implementations is opinionated about future version of the specification. Current
```
Comment (mtwo) [inline spec/20-http_header_format.md]: Implementations should be opinionated about future version of the specification. The current

Comment (SergeyKanzhelev) [reply spec/20-http_header_format.md]: nice catch! I did a poor find/replace job, but it caught bigger problem. I believe this sentence should be reworded as it talks about spec being opinionated, not implementation. Fixed.

**Review by mtwo — APPROVED** [review]
Reviewed at Sergey's request

**Review by AloisReitbauer — APPROVED** [review]

**Review by nikmd23 — APPROVED** [review]
LGTM :shipit:

**Pull request merged**